core/src/main/java/io/questdb/cairo/TableWriter.java

```diff
@@ -6143,12 +6185,15 @@ private void processO3Block(
                     .I$();
 
             o3ConsumePartitionUpdates();
+            if (o3ErrorCount.get() == 0) {
+                o3ConsumePartitionUpdateSink();
+            }
             o3DoneLatch.await(latchCount);
 
             o3InError = !success || o3ErrorCount.get() > 0;
             if (success && o3ErrorCount.get() > 0) {
                 //noinspection ThrowFromFinallyBlock
-                throw CairoException.critical(0).put("bulk update failed and will be rolled back");
+                throw CairoException.critical(0).put("bulk update failed and will be rolled back").setOutOfMemory(o3oomObserved);
```

Comment:
cc @mtopolnik

Reply:
I understood @mtopolnik and @ideoma agreed to changed CairoException in a subsequent PR since it's unrelated to this change. thus marking this as resolved.